### PR TITLE
Add LocallyConnected2DLayer, including documentation and tests.

### PR DESCRIPTION
**Pull request by alexlee-gk**
This PR follows up on the discussion from [here](https://groups.google.com/forum/#!topic/lasagne-users/G-LF7JCu9PM). I believe there is no issue relating to this though.

### CODE REVIEW

File: lasagne/layers/local.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+
+    num_filters : int
+        The number of learnable convolutional filters this layer has.
+
+    filter_size : int or iterable of int
+        An integer or a 2-element tuple specifying the size of the filters.
+
+    stride : int or iterable of int
+        An integer or a 2-element tuple specifying the stride of the
+        convolution operation.
```
Comment (f0k) [inline lasagne/layers/local.py]: Add a note that "This implementation only supports unit stride, the argument is provided for compatibility to convolutional layers only."

File: lasagne/layers/local.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+    def get_W_shape(self):
+        num_input_channels = self.input_shape[1]
+        output_shape = self.get_output_shape_for(self.input_shape)
+        if self.channelwise:
+            if self.channelwise and self.num_filters != num_input_channels:
+                raise ValueError("num_filters and the number of input "
+                                 "channels should match when channelwise is "
+                                 "true, but got num_filters=%r and %d input "
+                                 "channels" %
+                                 (self.num_filters, num_input_channels))
```
Comment (f0k) [inline lasagne/layers/local.py]: Hmm. Actually this should be in the constructor, but `self.input_shape` is only set when calling the super constructor, and the super constructor already calls `get_W_shape`. Probably it's okay here.

**Comment by benanne** [comment]
This looks neat, thanks for contributing :)

File: lasagne/layers/local.py (hunk 1 above)
Comment (alexlee-gk) [inline lasagne/layers/local.py]: Sure, I just updated the pull request to include this comment.

File: lasagne/layers/local.py (hunk 2 above)
Comment (alexlee-gk) [inline lasagne/layers/local.py]: Yeah, I thought about that. At least the exception is still raised during construction.

File: lasagne/layers/local.py [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+            return (self.num_filters,) + self.filter_size + output_shape[-2:]
+        else:
+            return (self.num_filters, num_input_channels) + \
+                   self.filter_size + output_shape[-2:]
+
+    def convolve(self, input, **kwargs):
+        output_shape = self.get_output_shape_for(self.input_shape)
```
Comment (f0k) [inline lasagne/layers/local.py]: `output_shape = self.output_shape` would be enough here -- we're past the constructor, so this has been set already. It was only required in `get_W_shape()`.

File: lasagne/layers/local.py [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+            raise NotImplementedError(
+                "LocallyConnected2DLayer requires stride=1 / (1, 1), but got "
+                "%r." % (stride,))
+        # require same convolution
+        if self.pad != 'same':
+            raise NotImplementedError(
+                "LocallyConnected2DLayer requires pad='same', but got %r." %
+                (pad,))
+
+    def get_W_shape(self):
```
Comment (f0k) [inline lasagne/layers/local.py]: Would be cool to have a check against variable input shape here. Similar to a dense layer, we require the input shape to be known in a locally-connected layer. Something like:
```python
if any(s is None for s in self.input_shape[1:]):
    raise ValueError(
            "A LocallyConnected2DLayer requires a fixed input shape "
            "(except for the batch size). Got %r." % (self.input_shape,))
```
Otherwise, there'll be an error in creating the weight matrix that's more difficult to interpret.

**Comment by f0k** [comment]
Thanks for fixing the docstring! I've had time to read through the implementation now, and have two minor comments -- if you don't want to address them, let me know and I'll address them myself.

**Comment by alexlee-gk** [comment]
Thanks for the feedback! I incorporated your suggestions and updated the PR.

File: lasagne/tests/layers/test_local.py [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            LocallyConnected2DLayer(input_layer, 2, 3, stride=2)
+        assert "requires stride=1 / (1, 1)" in exc.value.args[0]
+
+    def test_invalid_settings(self, DummyInputLayer):
+        from lasagne.layers import LocallyConnected2DLayer
+        input_layer = DummyInputLayer((10, 2, 4, 4))
+        with pytest.raises(ValueError) as exc:
+            LocallyConnected2DLayer(input_layer, 4, 3, channelwise=True)
+        assert "num_filters and the number of input channels should match"\
+               in exc.value.args[0]
```
Comment (f0k) [inline lasagne/tests/layers/test_local.py]: Thanks for your quick update! You'll need to add a test with `DummyInputLayer((10, None, 4, 4))` or something like this to trigger the new check for a fixed input size.

Comment (alexlee-gk) [inline lasagne/tests/layers/test_local.py]: Alright, just added a test for that. The PR passes all tests now.

**Comment by f0k** [comment]
Perfect! Thanks again for your clean contribution, finally merging!
